### PR TITLE
Create user map directory if it doesn't exist.

### LINKED ISSUE #13558: Crash when creating a new map when install dir is read-only
... and the support dir `maps` directory doesn't exist.  I believe this is a regression from #13223 fixing `Folder` spamming empty directories all over the place.

```
Exception of type `System.Reflection.TargetInvocationException`: Exception has been thrown by the target of an invocation.
Inner
  Exception of type `System.InvalidOperationException`: Sequence contains no elements
    at System.Linq.Enumerable.First[TSource] (System.Collections.Generic.IEnumerable`1[T] source) [0x00010] in <26aedeede9534b948c539f8734c8492d>:0 
  at OpenRA.Mods.Common.Widgets.Logic.SaveMapLogic..ctor (OpenRA.Widgets.Widget widget, OpenRA.ModData modData, System.Action`1[T] onSave, System.Action onExit, OpenRA.Map map, System.Collections.Generic.List`1[T] playerDefinitions, System.Collections.Generic.List`1[T] actorDefinitions) [0x00335] in /home/paul/src/OpenRA/OpenRA.Mods.Common/Widgets/Logic/Editor/SaveMapLogic.cs:119 
```

The simplest fix here is to add a `Directory.CreateDirectory` to the writeable-directory check in `SaveMapLogic`.  Ideally we would also show an error dialog instead of crashing if that for some reason *still* failed to find a usable save location.

This is an edge case so probably doesn't need to block our first playtest, but *does* need to be fixed before the final release.

### PR DESCRIPTION
**Pull request by pchote**
Fixes #13558.

### CODE REVIEW

File: OpenRA.Game/Map/MapCache.cs [hunk 1]
```diff
@@ -54,12 +54,20 @@ public MapCache(ModData modData)
...
 
 				IReadOnlyPackage package;
-				var optional = name.StartsWith("~");
+				var optional = name.StartsWith("~", StringComparison.Ordinal);
 				if (optional)
 					name = name.Substring(1);
 
 				try
 				{
+					// HACK: If this a reference to the support directory then we may need to create it
```
Comment (reaperrr) [inline OpenRA.Game/Map/MapCache.cs]: Nit: "If this *is* a reference (...)"

**Review by reaperrr — APPROVED** [review]

File: OpenRA.Game/Map/MapCache.cs [hunk 2]
```diff
@@ -54,12 +54,20 @@ public MapCache(ModData modData)
...
+					if (name.StartsWith("^", StringComparison.Ordinal))
+					{
+						var resolved = Platform.ResolvePath(name);
+						if (!File.Exists(resolved))
+							Directory.CreateDirectory(resolved);
```
Comment (Phrohdoh) [inline OpenRA.Game/Map/MapCache.cs]: You're looking for a file but create a dir?

Comment (pchote) [inline OpenRA.Game/Map/MapCache.cs]: Checking for `Directory.Exists` is redundant because CreateDirectory will not fail in that case.

Comment (Phrohdoh) [inline OpenRA.Game/Map/MapCache.cs]: So why do a check at all? Do we not know if this should be a file or a directory?

Comment (pchote) [inline OpenRA.Game/Map/MapCache.cs]: Indeed we do not.

**Comment by pchote** [comment]
Updated.

**Review by abcdefg30 — APPROVED** [review]

**Comment by abcdefg30** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/6948b4aa5b2a9a497fbfa8ca1b5c2b3eba86f10b...b56000899df615cc5a5712f06842b4a453beb022)

**Pull request merged**